perses/dispersed/feptasks.py

```diff
@@ -84,30 +110,97 @@ def _get_integrator(self, thermodynamic_state):
+        Parameters
+        ----------
+        thermodynamic_state : openmmtools.states.ThermodynamicState
+           The thermodynamic state to use to propagate dynamics.
+        sampler_state : openmmtools.states.SamplerState
+           The sampler state to apply the move to. This is modified.
+        See Also
+        --------
+        openmmtools.utils.Timer
+        """
+        # Check if we have to use the global cache.
+        if self.context_cache is None:
+            context_cache = cache.global_context_cache
+        else:
+            context_cache = self.context_cache
+
+        # Create integrator.
+        integrator = self._get_integrator(thermodynamic_state)
+
+        context, integrator = context_cache.get_context(thermodynamic_state, integrator)
+
+        # Subclasses may implement _before_integration().
+        self._before_integration(context, thermodynamic_state)
+        
+        self._cumulative_work[0] = integrator.get_protocol_work()
+
+        #loop through the number of times we have to apply in order to collect the requested work and trajectory statistics.
+        for iteration in range(self._number_of_step_moves):
+
+            integrator.step(self._number_of_step_moves)
+            self._current_total_work = integrator.get_protocol_work(dimensionless=True)
+            self._cumulative_work[iteration+1] = self._current_total_work
+
+            #if we have a trajectory, we'll also write to it
+            if self._topology is not None:
+                context_state = context.getState(getPositions=True, getVelocities=True, getEnergy=True,
+                                                 enforcePeriodicBox=thermodynamic_state.is_periodic)
+                sampler_state.update_from_context(context_state)
+                
+                #record positions for writing to trajectory
+                #we need to check whether the user has requested to subset atoms (excluding water, for instance)
+
+                if self._subset_atoms is None:
+                    self._trajectory_positions[iteration, :, :] = sampler_state.positions[:, :].value_in_unit_system(unit.md_unit_system)
+                else:
+                    self._trajectory_positions[iteration, :, :] = sampler_state.positions[self._subset_atoms, :].value_in_unit_system(unit.md_unit_system)
+
+                #get the box angles and lengths
+                a, b, c, alpha, beta, gamma = mdtrajutils.unitcell.box_vectors_to_lengths_and_angles(*sampler_state.box_vectors)
+                self._trajectory_box_lengths[iteration, :] = [a, b, c]
+                self._trajectory_box_angles[iteration, :] = [alpha, beta, gamma]
+
+
+        # We get also velocities here even if we don't need them because we
+        # will recycle this State to update the sampler state object. This
+        # way we won't need a second call to Context.getState().
+        context_state = context.getState(getPositions=True, getVelocities=True, getEnergy=True,
```

Comment:
This new `context_state` is not used anywhere. Do you want to update the MCMC sampler state from the current context instead?

Reply:
I use it in line 191.